From 5bc6095d3ea22028db9d6e50b6d33fb437001dcb Mon Sep 17 00:00:00 2001
From: Zhong Jinghua <zhongjinghua@huawei.com>
Date: Tue, 14 Mar 2023 15:53:30 +0800
Subject: [PATCH 0324/1646] [Huawei] Revert "[Huawei] block: fix null-deref in
 percpu_ref_put"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187268

----------------------------------------

This reverts commit 290245eda066c4d5f89cbe7c7e8f028fe2777141.

There is a new fix for this problem in the mainline patch, so the patch
should return to the mainline solution.

mainline patch:
d36a9ea5e776 ("block: fix use-after-free of q->q_usage_counter")

Fixes: 290245eda066("block: fix null-deref in percpu_ref_put")
Signed-off-by: Zhong Jinghua <zhongjinghua@huawei.com>
---
 block/blk-core.c       | 4 +---
 block/blk-mq.c         | 7 -------
 include/linux/blk-mq.h | 1 -
 include/linux/blkdev.h | 2 --
 4 files changed, 1 insertion(+), 13 deletions(-)

diff --git a/block/blk-core.c b/block/blk-core.c
index e653155952d3..70c304594faa 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -411,8 +411,7 @@ void blk_cleanup_queue(struct request_queue *q)
 	 * prevent that blk_mq_run_hw_queues() accesses the hardware queues
 	 * after draining finished.
 	 */
-	blk_freeze_queue_start(q);
-	blk_mq_freeze_queue_wait_sync(q);
+	blk_freeze_queue(q);
 
 	blk_queue_flag_set(QUEUE_FLAG_DEAD, q);
 
@@ -527,7 +526,6 @@ static void blk_queue_usage_counter_release(struct percpu_ref *ref)
 	struct request_queue *q =
 		container_of(ref, struct request_queue, q_usage_counter);
 
-	blk_queue_flag_set(QUEUE_FLAG_USAGE_COUNT_SYNC, q);
 	wake_up_all(&q->mq_freeze_wq);
 }
 
diff --git a/block/blk-mq.c b/block/blk-mq.c
index afc98a28f649..4fdb6fcc3790 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -195,7 +195,6 @@ void blk_freeze_queue_start(struct request_queue *q)
 {
 	mutex_lock(&q->mq_freeze_lock);
 	if (++q->mq_freeze_depth == 1) {
-		blk_queue_flag_clear(QUEUE_FLAG_USAGE_COUNT_SYNC, q);
 		percpu_ref_kill(&q->q_usage_counter);
 		mutex_unlock(&q->mq_freeze_lock);
 		if (queue_is_mq(q))
@@ -206,12 +205,6 @@ void blk_freeze_queue_start(struct request_queue *q)
 }
 EXPORT_SYMBOL_GPL(blk_freeze_queue_start);
 
-void blk_mq_freeze_queue_wait_sync(struct request_queue *q)
-{
-	wait_event(q->mq_freeze_wq, percpu_ref_is_zero(&q->q_usage_counter) &&
-			test_bit(QUEUE_FLAG_USAGE_COUNT_SYNC, &q->queue_flags));
-}
-
 void blk_mq_freeze_queue_wait(struct request_queue *q)
 {
 	wait_event(q->mq_freeze_wq, percpu_ref_is_zero(&q->q_usage_counter));
diff --git a/include/linux/blk-mq.h b/include/linux/blk-mq.h
index f3d78b939e01..2eca0b76b3e9 100644
--- a/include/linux/blk-mq.h
+++ b/include/linux/blk-mq.h
@@ -566,7 +566,6 @@ void blk_mq_freeze_queue(struct request_queue *q);
 void blk_mq_unfreeze_queue(struct request_queue *q);
 void blk_freeze_queue_start(struct request_queue *q);
 void blk_mq_freeze_queue_wait(struct request_queue *q);
-void blk_mq_freeze_queue_wait_sync(struct request_queue *q);
 int blk_mq_freeze_queue_wait_timeout(struct request_queue *q,
 				     unsigned long timeout);
 
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index b04613bc3ed5..9ede0a81e466 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -643,8 +643,6 @@ struct request_queue {
 #define QUEUE_FLAG_NOWAIT       29	/* device supports NOWAIT */
 /*at least one blk-mq hctx can't get driver tag */
 #define QUEUE_FLAG_HCTX_WAIT	30
-/* sync for q_usage_counter */
-#define QUEUE_FLAG_USAGE_COUNT_SYNC    31
 
 #define QUEUE_FLAG_MQ_DEFAULT	((1 << QUEUE_FLAG_IO_STAT) |		\
 				 (1 << QUEUE_FLAG_SAME_COMP) |		\
-- 
2.17.1

